include/stl2/detail/iterator/any_iterator.hpp

```diff
@@ -0,0 +1,230 @@
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/caseycarter/cmcstl2
+//
+#ifndef STL2_DETAIL_ITERATOR_ANY_ITERATOR_HPP
+#define STL2_DETAIL_ITERATOR_ANY_ITERATOR_HPP
+
+#include <new>
+#include <atomic>
+#include <stl2/type_traits.hpp>
+#include <stl2/detail/fwd.hpp>
+#include <stl2/detail/swap.hpp>
+#include <stl2/detail/concepts/core.hpp>
+#include <stl2/detail/iterator/basic_iterator.hpp>
+#include <stl2/detail/iterator/concepts.hpp>
+
+STL2_OPEN_NAMESPACE {
+	namespace __any_iterator {
+        enum class op { copy, move, nuke, bump, comp, rval };
+
+        struct counted {
+            std::atomic<long> cnt{ 1 };
+        };
+        InputIterator{I} struct shared_iterator : counted {
+            shared_iterator(I i) : it(std::move(i)) {}
+            I it;
+        };
+
+        union blob {
+            counted* big;
+            std::aligned_storage_t<2 * sizeof(void *)> tiny;
+        };
+
+        template <class It>
+        using is_small =
+            std::integral_constant<bool, (sizeof(It) <= sizeof(blob::tiny))>;
+        using small_tag = std::true_type;
+        using big_tag = std::false_type;
+
+        template <class RValueReference>
+        using iter_move_fn = RValueReference (*)(blob const &);
+
+        template <class RValueReference>
+        inline iter_move_fn<RValueReference> uninit_noop(op, blob *, blob *) {
+            return nullptr;
+        }
+
+        template <class Reference>
+        [[noreturn]] inline Reference uninit_deref(blob const &) {
+            std::terminate();
+        }
+
+        template <class Reference, InputIterator I>
+        static Reference deref_small(blob const &src) {
+            return **static_cast<I const *>(static_cast<void const *>(&src.tiny));
+        }
+
+        template <class Reference, InputIterator I>
+        static Reference deref_big(blob const &src) {
+            return *static_cast<shared_iterator<I> const *>(src.big)->it;
+        }
+
+        template <class RValueReference, InputIterator I>
+        iter_move_fn<RValueReference> exec_small(op o, blob *src, blob *dst) {
+            switch (o) {
+            case op::copy:
+                ::new (static_cast<void *>(&dst->tiny))
+                    I(*static_cast<I const *>(static_cast<void *>(&src->tiny)));
+                break;
+            case op::move:
+                ::new (static_cast<void *>(&dst->tiny))
+                    I(std::move(*static_cast<I *>(static_cast<void *>(&src->tiny))));
+                // fallthrough
+            case op::nuke:
+                static_cast<I *>(static_cast<void *>(&src->tiny))->~I();
+                break;
+            case op::bump:
+                ++*static_cast<I *>(static_cast<void *>(&src->tiny));
+                break;
+            case op::comp:
+                if (*static_cast<I const *>(static_cast<void *>(&src->tiny)) ==
```

Comment:
... or `==` could always return true when `I` doesn't satisfy `Sentinel<I, I>()`; though that would be easier implemented with a concrete vtable than this vtable-in-a-function.